bugfix/ATC-2

erikquinn opened the pull request:
Honestly, I've always been confused as to why this bit of code worked before, because it didn't make sense.... apparently something you changed made this wrong logic start being, well.. wrong! lol

Nonetheless, here's part of the resolution to #2. Might be more to it, but I'm done for the day, but wanted to post this here at the very least for now. Probably some issues beyond just this, hence the WIP label.

erikquinn committed:
Get ILS intercepting again

erikquinn committed:
Fix altitude behavior during ILS interception

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2021,7 +2021,7 @@ export default class Aircraft {
-            if ((abs(this.altitude - glideslope_altitude) < glideslope_window)
+            if (this.altitude < (glideslope_altitude + glideslope_window)
                 && (abs(offset_angle) < degreesToRadians(10))
                 && (offset[1] < ils)
```
Something I would like to see the codebase moving towards is abstraction of these large logic blocks.

``` javascript
canLockILS() {
    return this.altitude < (glideslope_altitude + glideslope_window)
        && (abs(offset_angle) < degreesToRadians(10))
        && (offset[1] < ils);
}
```

which can then be used as:

``` javascript
if (this.canLockILS()) {
    ...
}
```

This is obviously just an example and in reality, the above abstraction should go another step further to be `isWithinAngle()` and `isWithinAltitude()` which would then become

``` javascript
canLockILS() {
    return this.isWithinAngle() && this.isWIthinAltitude();
}
```

Even though it seems like more code, and it is, is is much more readable.  It is also more portable.  It is also easier to test.

n8rzz replied:
Also, not something that needs to be fixed now as these large logic blocks are everywhere.  But something to keep in mind for the future.

erikquinn replied:
Agreed. I actually originally wanted to take it even further, and abstract it to the level of intercepting a lateral course, since this has countless applications (intercepting localizers, joining airways, VOR radials, holdings, even taxi paths, etc).

n8rzz replied:
I like where your head is at!  I just wanted to call that out for future work.  These will probably need to be tackled as-we-come-to-them.

n8rzz commented:
@erikquinn are you still working on this one?  Or should I take a look at this and get it merged with the partial fix you have in place?

erikquinn commented:
@n8rzz sorry for the delay, I'd like to have a crack at fixing the goofiness that sometimes comes up with localizer interception while we're here. I've got some time now, if I get it fixed I'll add it in, and if not, what's already here at least solves the immediate issue.

n8rzz commented:
no worries.  go ahead and take a crack at it!

erikquinn committed:
Improve localizer capture behavior

erikquinn commented:
Localizer funny business removed. My god that was painful... lol

Review at will!

n8rzz commented on src/assets/scripts/math/circle.js:
```diff
@@ -64,3 +64,27 @@ export const fix_angle = (radians) => {
+    if (radians < 0) {
+        return (Math.PI*2) + (radians % (Math.PI*2));
+    }
+    else return radians % (Math.PI*2);
```
We have `tau()` available for the `Math.PI * 2` calculations.  I would like to see this logic flipped to remove the `else`.  
instead of:

``` javascript
if (radians < 0) {
    ...
}
```

flip it to:

``` javascript
if (radians >= 0) {
    return radians % tau();
}

return tau() + (radians % tau());
```

This way we can remove the else and make the function easier to read.  Also, for readability, please do not use one liners with if/else statements.

erikquinn replied:
With `tau` being in the same file as an _export_, and not as a variable, my attempts to call it were coming up undefined... internet said to either import the file to itself or to access it via `module.exports.tau()`, but these are being exported via the `export` keyword directly, and not as `module.exports`, so I figured it wouldn't work to do that either (admittedly I didn't try). I fully intended to use tau elsewhere though, just had trouble using it in the same file.

And yeah it can be flopped, I care not. haha as for the oneline if/else, that's fine by me. I'll remove it.

n8rzz replied:
ahh exports.  When I first started using es2015 the whole import/export thing had me stumped more than once.  The syntax was goofy and I didn't understand what the differences were.

``` javascript
// default export
// Foo.js
export default class Cube {
   ...
}

// doesn't need curly braces when importing.  Typically default exports are reserved for classes or when a file has a single export.
// bar.js
import Cube from './Foo';

// regular export
// wow.js
export const SOME_CONSTANT = 'A String!';

export const calculateSomething = () => {
    ....
};

// or without the arrow function (which brings in scoping concerns with `this`
export const calculateSomethingElse = function calculateSomethingElse() {
    ....
}

// zoo.js
import { SOME_CONSTANT, calculateSomething, calculateSomethingElse } from './wow';
```

When you are dealing with exported function within the same file, you can use them just as would would normally.

``` javascript
export const something = () => {
    ...
};

export const somethingThatDoesSomethingElse = () => {
    const value = something();
}
```

There is one other type, the wildcard import, but that one can have side effects and is inefficient.  The two examples above are what you will see throughout this project.

erikquinn replied:
If the latter is true, then I must have just done something stupid like spelled it wrong, or forgot to re-transpile... oops! I'll put it back to calling `tau()` if it does just work.

n8rzz replied:
spelling is hard.  9 times out of 10 it is spelling that causes my bugs/breakages.

n8rzz commented on src/assets/scripts/math/circle.js:
```diff
@@ -64,3 +64,27 @@ export const fix_angle = (radians) => {
+};
+
+/**
+ * normalize angles to within 0 - 2π
```
thank you for adding a comment!!

erikquinn replied:
Always!

I am curious though, why not have it set up like the below, so when collapsing code to the level of those functions/methods/etc you can SEE the brief description. You may have come across some commenting done like the below, which was my variant. It clearly doesn't seem to be the way it's done by the pros though-- I'm curious as to why?

``` javascript
/** normalize angles to within 0° - 360°
 ** @param  {number} degrees an angle
 ** @return {number}         an angle within [0,360]
 */
```

n8rzz replied:
the main reason is because [YuiDoc](http://yui.github.io/yuidoc/syntax/index.html) and [JSDOC](http://usejsdoc.org/howto-es2015-classes.html).  Both of these things output api documentation based on comment blocks.  See [StrutureJs](http://codebelt.github.io/StructureJS/docs/) written by some people I work with, for an example of what that actually means.  All the documentation there was generated from comment blocks, just like the ones I'm using here.

I tend to lean more towards the YuiDoc syntaxes, but they are mostly the same.

n8rzz replied:
Maybe someday I'll add a doc gen task that will generate proper documentation for this.  Thats definitely a longterm goal for me.

n8rzz replied:
just to clarify, both of those doc gens actually _generate_ html pages based on the doc blocks within the code.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2066,7 +2078,8 @@ export default class Aircraft {
 
                 this.target.speed = crange(3, offset[1], 10, this.model.speed.landing, this.fms.currentWaypoint().start_speed);
-            } else if ((this.altitude - runway_elevation) >= 300 && this.mode === FLIGHT_MODES.LANDING) {
+            }
```
make sure an `else if` is on the same line as the closing brace for the previous conditional.

``` javascript
// do
if (condition) {
    ...
} else if (otherCondition) {
    ...
}

// do not
if (condition) {
    ...
} 
else if (otherCondition) {
    ...
}

```

erikquinn replied:
I was explicitly changing this to that. I consider it to be way less readable when I don't see an `else if` in vertical alignment with its own closing bracket. eg seeing "if" --> "{" --> "{" --> "{" when I look down a column seems to me to be less readable.

Again, I can change (I'm actually enjoying learning all this), just am curious as to the reasoning here.

n8rzz replied:
There are two points of emphasis here:
- I am, again, following company standards here https://github.com/thenerdery/javascript-standards#blocks with block statements.  In C# you will see lines broken up similarly to what you have there.  General javascript consensus (that I've seen anyway) tends to lean towards same line braces `} else if () {`
- I would argue that there should rarely be a need for an `else if`.  There are always exceptions, of course, but I am of the thought that `else if`s are generally a sign of a code smell.

That last one may take a minute to sink in.  You may ask yourself, "How would I do large logic blocks without `else if`?"  Abstraction, helper methods, cleaner logic.

---

I'm working on some `v3.0.x` stuff that I haven't pushed yet with the `AirportModel` and others that all started with this block here: https://github.com/n8rzz/atc/blob/develop/src/assets/scripts/airport/AirportModel.js#L557

When I _do_ push it you may look at it and think that I haven't simplified anything at all.  In fact I've made things much, much more complicated with 5 new files and a whole new set of class methods.  At first glance, it will undoubtedly _look_ more complicated, but I can tell you, it makes things a whole lot simpler.  It will take a little time to get used to it.  

That whole block in the `AirportModel` from above, that will instead become:

``` javascript
getSID(id, exit, rwy) {
    return this.sidCollection.findFixesForSidByRunwayAndExit(id, exit, rwy);
}
```

This then fires off a chain of collections and models that organize and provide methods for reasoning about a SID and it's fixes.  Oh, and all these new classes and methods are fully tested :-)  So I know what these methods will do under certain circumstances.

Anyway, this is for the future.  We can talk more about simplification then, after `v3.0.0`.

erikquinn replied:
This gives me the heebiejeebies, just because it's dramatically different, but I could see how it could be much easier to work with. #inNateWeTrust

n8rzz replied:
It is a little scary, that's normal.  I've been there.  I can remember going over a simple thing with my mentor a while back, and we went down this same path where a 100(ish) line block of code became a whole bunch of files.  I asked, 'How the hell is this better!?'  

You'll see.  I might just do these changes as a PR so we can talk about it.  I fully expect conversations about the whys and the 'I dont get what this is doing', or just plain 'Whaaaaat??'.  Totally normal, been there.  It _will_ end up being not-so-scary, I promise.

I've said it before, but just to reiterate, _please_ ask questions.  You're doing a wonderful job of that so far.  Continue that.  Ask when you see something you don't understand or when you have questions.  That's how we all get better as developers. 

Hmm, I am both encouraged and terrified by that hashtag...

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2080,8 +2093,9 @@ export default class Aircraft {
-            } else if (this.altitude >= 300) {
-                this.target.heading = this.fms.currentWaypoint().heading;
+            }
+            else if (this.altitude >= 300) {
```
make sure an else if is on the same line as the closing brace for the previous conditional.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2091,7 +2105,9 @@ export default class Aircraft {
-        } else if (this.fms.currentWaypoint().navmode === WAYPOINT_NAV_MODE.FIX) {
+        }
+
+        else if (this.fms.currentWaypoint().navmode === WAYPOINT_NAV_MODE.FIX) {
```
make sure an else if is on the same line as the closing brace for the previous conditional.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2118,7 +2134,9 @@ export default class Aircraft {
-        } else if (this.fms.currentWaypoint().navmode === WAYPOINT_NAV_MODE.HOLD) {
+        }
+
+        else if (this.fms.currentWaypoint().navmode === WAYPOINT_NAV_MODE.HOLD) {
```
make sure an else if is on the same line as the closing brace for the previous conditional.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2149,7 +2167,9 @@ export default class Aircraft {
-        } else {
+        }
+
+        else {
```
make sure an else if is on the same line as the closing brace for the previous conditional.

n8rzz commented on src/assets/scripts/airport/Runway.js:
```diff
@@ -97,7 +98,7 @@ export default class Runway {
             // TODO: what is the 0.5 for? enumerate the magic number
             this.midfield = vscale(vadd(thisSide.position, farSide.position), 0.5);
-            this.angle = vradial(vsub(farSide.position, thisSide.position));
+            this.angle = radians_normalize(vradial(vsub(farSide.position, thisSide.position)));
```
This should work for now, but I feel like this will need to be abstracted in `v3.0.x`.  What _exactly_ is this line doing; maybe something like `this.calculateRunwayAngle()`?

Doesn't have to be done now, but something to keep an eye out for.

erikquinn replied:
yeah, that would be better, for sure.

n8rzz reviewed (commented):
One major thing to change, with `tau()` and a logic change, then also a few nit picks on syntax.

One thing to keep in mind is that the `AirportInstanceModel` will likely get split up in `v.3.0.x`.  A colleague once had a rule he called 'The Middle Finger Rule'; if a method is longer that your middle finger, its too long.  This file has _a lot_ of those methods here.  2700+ lines is aay too long.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2048,16 +2048,28 @@ export default class Aircraft {
                     if (offset[1] > 0.2 && abs(offset[0]) > 0.003) {
-                        // TODO: enumerate the magic numbers
-                        this.target.heading = _clamp(degreesToRadians(-30), -12 * offset_angle, degreesToRadians(30)) + angle;
+                        const severity_of_correction = 12;  // controls steepness of heading adjustments during localizer tracking
```
👍 for enumeration!

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2048,16 +2048,28 @@ export default class Aircraft {
+                        let minHdg = null;
+                        let maxHdg = null;
+                        if (!assignedHdg) {  // interception via fixes
+                            // TODO: uh.... do this.
```
so what happens if we get inside this block; does this cause an error?  is it even possible to get here?  does this break everything and cause the world to end?

erikquinn replied:
yeah it's possible to get in there. People using fixes to put a/c on the loc WOULD get there. And here it would just fail to intercept, which I guess a silent failure is a doofy solution. I intend to get some full logic in there, but I suppose it should at least have had a warning or error of some sort so when it doesn't work, it is clear why.

n8rzz replied:
One thing my mentors always ask ask is who is the audience for this error?
- Does the user need to know about it?
- Or is this meant for developers?

Once you figure out the audience, then you need to decide on how to present it to that audience.  

In this instance, I would probably go with a `console.warn()` with a short message.  Will an error here actually break anything, like cause other parts of the app to fail?  Or will the app keep plugging away?

I'd also fill out that `TODO` block with a little more detail about what needs to be done.  Doesn't need to be a book, just a quick note so the next person that deals with it knows why there is an empty if block and what the future may hold for said block.

erikquinn replied:
It'll continue running, it just won't capture the ILS, despite having been cleared for the approach.

n8rzz replied:
right on.  

this sounds like a good place for a comment and a `console.warn`, for the developers out there.

n8rzz commented:
retargeting this PR to `release/3.0.0` (should have done this from the beginning).  `develop` will now be used for v3.0.x commits.

n8rzz commented:
also, new bugfix branches should be branching off of `release/3.0.0` now instead of develop

erikquinn committed:
Rename and document confusing crange/trange functions

erikquinn committed:
Code cleanup and improvements to ILS capture logic
address comments from code review

erikquinn commented:
Pushed some more changes to this:
- Addressed code review comments
- cleaned that whole section up a little more (it's still not so good, but a little easier to follow now)
- fixed a couple minor weirdnesses with localizer interception
- added a snitch on glideslope height
- figured out `crange` and `trange` functions, documented them, and fixed a mistake in how they were being fed parameters that was leading to weirdness with speed control on final
- rebased onto the latest revision of `release/3.0.0`

n8rzz commented:
Awesome!! I'll take a look at this once I get home tonight.

erikquinn commented:
So this whole `_clamp` situation is still kinda bewildering, but I've been able to pin down a reproducible way to see the issue and not see the issue. Maybe you can deduce the cause from it, because I'm at a loss for why...

The following function does the appropriate math, but causes the targets to be jumpy on radar:

``` javascript
export const extrapolate_range_clamp = (range1_min, target_val, range1_max, range2_min, range2_max) => {
    const extrapolation_result = extrapolate_range(range1_min, target_val, range1_max, range2_min, range2_max);
    return _clamp(extrapolation_result, range2_min, range2_max);
};

// only the last line differs from the other example
```

---

Rearranging the arguments yields a useless result, but the targets move smoothly:

``` javascript
export const extrapolate_range_clamp = (range1_min, target_val, range1_max, range2_min, range2_max) => {
    const extrapolation_result = extrapolate_range(range1_min, target_val, range1_max, range2_min, range2_max);
    return _clamp(range2_min, extrapolation_result, range2_max);
};

// only the last line differs from the other example
```

---

If you check out this pull, and reorder the arguments at `src/assets/scripts/math/core.js:142`, you can observe the difference when running the sim at 150x rate via `timewarp 150`.

n8rzz commented:
Have you tried it with our custom `clamp()` function?  And with the same results?

n8rzz commented:
`_clamp()` does more logic than is probably necessary in our case.  it would make sense for there to be sluggishness.  see:
https://github.com/lodash/lodash/blob/4.16.4/lodash.js#L13859

erikquinn commented:
Was able to get it working with the custom clamper without any issues. Didn't add them here, since they are already in another pull soon go to in.

This PR is complete, but depends on #8.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -1995,101 +1994,115 @@ export default class Aircraft {
+                if (this.fms.currentWaypoint().speed > 0) {
+                    this.fms.setCurrent({ start_speed: this.fms.currentWaypoint().speed });
+                }
+                if (this.isLanded()) {
```
be sure to separate new conditional block logic by a blank line

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -1995,101 +1994,115 @@ export default class Aircraft {
+                const turn_early_km = 1;    // start turn 1km early, to avoid overshoots from tailwind
+                const should_attempt_intercept = (0 < dist_to_localizer && dist_to_localizer <= turning_radius + turn_early_km);
+                const in_the_window = abs(offset_angle) < degreesToRadians(1.5);  // if true, aircraft will move to localizer, regardless of assigned heading
+                if (should_attempt_intercept || in_the_window) {  // time to begin turn
```
be sure to separate new `if/else` block logic by a blank line

n8rzz commented on src/assets/scripts/airport/AirportModel.js:
```diff
@@ -95,8 +95,8 @@ export default class AirportModel {
-        wind.angle += crange(-1, angle_factor, 1, degreesToRadians(-4), degreesToRadians(4));
-        wind.speed *= crange(-1, speed_factor, 1, 0.9, 1.05);
+        wind.angle += extrapolate_range_clamp(-1, angle_factor, 1, degreesToRadians(-4), degreesToRadians(4));
+        wind.speed *= extrapolate_range_clamp(-1, speed_factor, 1, 0.9, 1.05);
```
what do all these numbers mean?  In the future I'd like to see these enumerated. Math is hard and I, or anybody else that may want to contribute, may not be able to figure out what each number is for or why a specific value is used.

n8rzz commented on src/assets/scripts/math/circle.js:
```diff
@@ -47,20 +47,25 @@ export const angle_offset = (a, b) => {
+ * @return {number}         an angle within [0,360]
  */
-export const fix_angle = (radians) => {
-    while (radians > tau()) {
```
+1 for getting rid of this while loop!!

n8rzz commented on src/assets/scripts/math/core.js:
```diff
@@ -108,35 +107,35 @@ export const mod = (firstValue, secondValue) => {
+ * @param  {number} range1_max maximum value of range 1
+ * @param  {number} range2_min minimum value of range 2
+ * @param  {number} range2_max maximum value of range 2
+ * @return {number}            target value wihtin range 2
```
+1 for adding comments and documenting what this is and what these values are.

n8rzz commented on src/assets/scripts/math/core.js:
```diff
@@ -108,35 +107,35 @@ export const mod = (firstValue, secondValue) => {
+ * @param  {number} range1_max maximum value of range1
+ * @param  {number} range2_min minimum value of range2
+ * @param  {number} range2_max maximum value of range2
+ * @return {number}            target value within range2
```
+1 here too, nice work!

n8rzz reviewed (commented):
Very nice.  I called out a few things, nothing that needs to be addressed here.  As usual, just some things to keep in mind for the future.  

I'm just resolving merge conflicts with this branch now (because of #1) and will then begin testing the clamps.  Stay tuned...

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -1995,101 +1994,115 @@ export default class Aircraft {
             offset = getOffset(this, runway.position, runway.angle);
             offset_angle = vradial(offset);
+            angle = radians_normalize(runway.angle);
+            glideslope_altitude = _clamp(0, runway.getGlideslopeAltitude(offset[1]), this.altitude);
```
fixed this `_clamp` import within merge conflicts

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -1995,101 +1994,115 @@ export default class Aircraft {
+                    const dist_assigned_spd = 9.5;  // 9.5km ~= 5nm
+                    this.target.speed = extrapolate_range_clamp(dist_final_app_spd, offset[1], dist_assigned_spd,
+                                            this.model.speed.landing, this.fms.currentWaypoint().start_speed);
+                    if(!this.projected) console.log(`start_speed:${this.fms.currentWaypoint().start_speed}  |  targetSpd:${this.target.speed}`);
```
be sure to remove dev notes once you're done with them

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -1995,101 +1994,115 @@ export default class Aircraft {
+                    const dist_final_app_spd = 3.5; // 3.5km ~= 2nm
+                    const dist_assigned_spd = 9.5;  // 9.5km ~= 5nm
+                    this.target.speed = extrapolate_range_clamp(dist_final_app_spd, offset[1], dist_assigned_spd,
+                                            this.model.speed.landing, this.fms.currentWaypoint().start_speed);
```
when parameters make a function/method call too long, separate the parameters on a single line:

``` javascript
this.target.speed = extrapolate_range_clamp(
    dist_final_app_spd, offset[1],
    dist_assigned_spd,
    this.model.speed.landing,
    this.fms.currentWaypoint().start_speed
);
```

n8rzz commented:
whoa.  I see exactly what you mean with `clamp` vs `_clamp` and it is _drastic_.

I think part of the problem there may be when the data is updated and when it is rendered.  There is no doubt the lodash function here is slower, but I think when those calculations are happening is as part of it.

Thats a problem to address at a later date: moving data updates to and render updates to their own loops.

n8rzz committed:
Merge branch 'release/3.0.0' of https://github.com/n8rzz/atc into bugfix/ATC-2
